src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,407 @@
+			configsBuilder.leaseManagementConfig(),
+			configsBuilder.lifecycleConfig(),
+			configsBuilder.metricsConfig(),
+			configsBuilder.processorConfig(),
+			configsBuilder.retrievalConfig());
+	}
+
+	@Override
+	protected void doStart() {
+		super.doStart();
+		this.executor.execute(() -> this.scheduler.run());
+	}
+
+	/**
+	 * Takes no action by default. Subclasses may override this if they need
+	 * lifecycle-managed behavior.
+	 */
+	@Override
+	protected void doStop() {
+
+		super.doStop();
+		this.scheduler.shutdown();
+
+	}
+
+	@Override
+	protected AttributeAccessor getErrorMessageAttributes(org.springframework.messaging.Message<?> message) {
+		AttributeAccessor attributes = attributesHolder.get();
+		if (attributes == null) {
+			return super.getErrorMessageAttributes(message);
+		}
+		else {
+			return attributes;
+		}
+	}
+
+	public void setStreamInitialSequence(InitialPositionInStream streamInitialSequence) {
+		setStreamInitialSequenceExtended(InitialPositionInStreamExtended.newInitialPosition(streamInitialSequence));
+	}
+
+	public void setStreamInitialSequenceExtended(InitialPositionInStreamExtended streamInitialSequence) {
+		Assert.notNull(streamInitialSequence, "'streamInitialSequence' must not be null");
+		this.streamInitialSequence = streamInitialSequence;
+	}
+
+	public void setIdleBetweenPolls(int idleBetweenPolls) {
+		this.idleBetweenPolls = Math.max(250, idleBetweenPolls);
+	}
+
+	public void setConsumerBackoff(int consumerBackoff) {
+		this.consumerBackoff = Math.max(1000, consumerBackoff);
+	}
+
+	/**
+	 * Sets the interval between 2 checkpoints.
+	 *
+	 * @param checkpointsInterval interval between 2 checkpoints (in milliseconds)
+	 */
+	public void setCheckpointsInterval(long checkpointsInterval) {
+		this.checkpointsInterval = checkpointsInterval;
+	}
+
+	@Override
+	public String toString() {
+		return "KclMessageDrivenChannelAdapter{consumerGroup='" + this.consumerGroup + '\'' + ", stream='" + this.stream + "'}";
+	}
+
+	private class RecordProcessorFactory implements ShardRecordProcessorFactory {
+		@Override
+		public ShardRecordProcessor shardRecordProcessor() {
+			return new RecordProcessor();
+		}
+	}
+
+	/**
+	 * Processes records and checkpoints progress.
+	 */
+	private class RecordProcessor implements ShardRecordProcessor {
+
+		private String shardId;
+
+		private long nextCheckpointTimeInMillis;
+
+		/** {@inheritDoc} */
+		@Override
+		public void initialize(InitializationInput initializationInput) {
+			this.shardId = initializationInput.shardId();
+			if (logger.isInfoEnabled()) {
+				logger.info("Initializing record processor for shard: " + this.shardId);
+			}
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void leaseLost(LeaseLostInput leaseLostInput) {
+			logger.info("Lost lease, so terminating.");
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void shardEnded(ShardEndedInput shardEndedInput) {
+			try {
+				logger.info("Reached shard end checkpointing.");
```

Comment:
I wonder if need to introduce something like `KinesisShardEndedEvent` and emit it from here.
There might be some logic in the application to react respectively.

Right, separate issue.